Las/Utils/Authorization.cs

```diff
@@ -0,0 +1,177 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+using Lucidtech.Las.Core;
+
+namespace Lucidtech.Las.Utils
+{
+    public class AmazonAuthorization
+    {
+        private string Region { get; }
+        private string Service { get; }
+        private Credentials Creds { get; }
+        private const string Algorithm = "AWS4-HMAC-SHA256";
+
+        public AmazonAuthorization(Credentials credentials)
+        {
+            Creds = credentials;
+            Region = "eu-west-1";
+            Service = "execute-api";
+        }
+
+        public Dictionary<string, string> SignHeaders(Uri uri, string method, byte[] body)
+        {
+            if (body == null)
+            {
+                body = Encoding.UTF8.GetBytes("");
+            }
+            
+            string amzDate = AmzDate();
+            string dateStamp = Datestamp();
+
+            Dictionary<string, string> headers = Headers(uri, amzDate);
+            
+            byte[] canonicalRequest = GetCanonicalRequest(uri, method, body, headers);
+            
+            string reqDigest = ComputeHashSha256(canonicalRequest);
+            string credScope = GetCredentialScope(dateStamp);
+            byte[] stringToSign = GetSignString(amzDate, credScope, reqDigest);
+            byte[] signingKey = GetSignatureKey(dateStamp);
+            string signature = StringFromByteArray(SignHash(signingKey, stringToSign));
+                 
+            var authHeader = BuildAuthHeader(
+                amzDate: amzDate,
+                signature: signature,
+                credScope: credScope,
+                signedHeaders: headers.Keys.ToList()
+            );
+            return authHeader;
+        }
+        
+        public static string StringFromByteArray(byte[] bytes)
+        {
+            var builder = new StringBuilder();
+            foreach (var b in bytes)
+            {
+                builder.Append(b.ToString("x2"));  
+            }
+            return builder.ToString();  
+        }
+        
+        public static byte[] SignHash(byte[] key, byte[] msg)
+        {
+            var hmac = new HMACSHA256(key);
+            return hmac.ComputeHash(msg);
+        }
+        
+        private byte[] GetCanonicalRequest(Uri uri, string method, byte[] body, Dictionary<string, string> headers)
+        {
+            string headerList = string.Join(";", headers.Keys);
+            
+            string headerParts = "";
+            foreach (var entry in headers)
+            {
+                headerParts = headerParts + $"{entry.Key}:{entry.Value}\n";
+            }
+
+            string payloadHash = ComputeHashSha256(body);
+            
+            var requestComponents = new List<string>() {
+                method,
+                uri.AbsolutePath,
+                uri.Query,
```

Comment:
Does uri.Query return the AWS canonical querystring? If not, would it be an idea to throw an exception if uri.Query is not the empty string to avoid getting a random signature calculation error upon sending the request? Alternatively implement  calculation of AWS canonical querystring